core/distributed/neighborhood_communicator.cpp

```diff
@@ -0,0 +1,273 @@
+        recv_buffer, recv_sizes_.data(), recv_offsets_.data(), recv_type,
+        comm_.get(), req.get()));
+    return req;
+#endif
+}
+
+
+std::unique_ptr<CollectiveCommunicator>
+NeighborhoodCommunicator::create_with_same_type(
+    communicator base, const distributed::index_map_variant& imap) const
+{
+    return std::visit(
+        [base](const auto& imap) {
+            return std::unique_ptr<CollectiveCommunicator>(
+                new NeighborhoodCommunicator(base, imap));
+        },
+        imap);
+}
+
+
+NeighborhoodCommunicator::NeighborhoodCommunicator(
+    NeighborhoodCommunicator&& other)
+    : NeighborhoodCommunicator(other.get_base_communicator())
+{
+    *this = std::move(other);
+}
+
+
+NeighborhoodCommunicator& NeighborhoodCommunicator::operator=(
+    NeighborhoodCommunicator&& other)
+{
+    if (this != &other) {
+        comm_ = std::exchange(other.comm_, MPI_COMM_NULL);
+        send_sizes_ =
+            std::exchange(other.send_sizes_, std::vector<comm_index_type>{});
+        send_offsets_ =
+            std::exchange(other.send_offsets_, std::vector<comm_index_type>{0});
+        recv_sizes_ =
+            std::exchange(other.recv_sizes_, std::vector<comm_index_type>{});
+        recv_offsets_ =
+            std::exchange(other.recv_offsets_, std::vector<comm_index_type>{0});
+    }
+    return *this;
+}
+
+
+bool operator==(const NeighborhoodCommunicator& a,
+                const NeighborhoodCommunicator& b)
+{
+    return (a.comm_.is_identical(b.comm_) || a.comm_.is_congruent(b.comm_)) &&
+           a.send_sizes_ == b.send_sizes_ && a.recv_sizes_ == b.recv_sizes_ &&
+           a.send_offsets_ == b.send_offsets_ &&
+           a.recv_offsets_ == b.recv_offsets_;
+}
+
+
+bool operator!=(const NeighborhoodCommunicator& a,
+                const NeighborhoodCommunicator& b)
+{
+    return !(a == b);
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+NeighborhoodCommunicator::NeighborhoodCommunicator(
+    communicator base,
+    const distributed::index_map<LocalIndexType, GlobalIndexType>& imap)
+    : CollectiveCommunicator(base),
+      comm_(MPI_COMM_NULL),
+      recv_sizes_(imap.get_remote_target_ids().get_size()),
+      recv_offsets_(recv_sizes_.size() + 1),
+      send_offsets_(1)
+{
+    auto exec = imap.get_executor();
+    if (!exec) {
+        return;
+    }
+    auto host_exec = exec->get_master();
+
+    auto recv_target_ids_arr =
+        make_temporary_clone(host_exec, &imap.get_remote_target_ids());
+    auto remote_idx_offsets_arr = make_temporary_clone(
+        host_exec, &imap.get_remote_global_idxs().get_offsets());
+    std::vector<comm_index_type> recv_target_ids(
+        recv_target_ids_arr->get_size());
+    std::copy_n(recv_target_ids_arr->get_const_data(),
+                recv_target_ids_arr->get_size(), recv_target_ids.begin());
+    for (size_type seg_id = 0;
+         seg_id < imap.get_remote_global_idxs().get_segment_count(); ++seg_id) {
```

Comment:
dump question: will `imap.get_remote_global_idxs().get_segment_count()` == `imap.get_remote_target_ids().get_size()`?

Reply:
Yes. IMO we could have a better data structure for this, because the `remote_target_ids` act like keys to the segments.